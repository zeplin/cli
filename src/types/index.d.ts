export * from "../commands/connect/interfaces/prism";
export * from "../commands/connect/interfaces/plugin";
export * from "../commands/connect/interfaces/config";
