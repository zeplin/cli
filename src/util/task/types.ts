import { Options as OraOptions } from "ora";
import { Task } from "./task";

export enum State {
    PENDING,
    COMPLETED,
    FAILED,
    SKIPPED
}

export type stringFn<T> = (ctx: T) => string;

export interface TaskUI<T = TaskContext> {
    readonly text?: string | stringFn<T>;
    readonly subtext?: string | stringFn<T>;
}

export type TaskUIFn<T> = stringFn<T> | ((context: T) => TaskUI<T>);

export type TaskUITypes<T> = void | string | TaskUI<T> | TaskUIFn<T>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type TaskContext = any;

export type TaskStep<T = TaskContext> = (ctx: T, task: Task<T>) => TaskUITypes<T> | Promise<TaskUITypes<T>>;

export type TaskSkipFunction<T> = (ctx: T) => boolean | TaskUITypes<T> | Promise<boolean | TaskUITypes<T>>;

export interface TaskConstructor<T> {
    readonly steps: TaskStep<T>[];
    readonly initial?: TaskUITypes<T>;
    readonly skip?: TaskSkipFunction<T>;
    readonly spinnerOptions?: OraOptions;
}