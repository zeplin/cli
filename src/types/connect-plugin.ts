import { PrismLang } from "./prism";
import { ComponentConfigFile } from "../commands/connect/interfaces";

export type UrlPath = { [keys: string]: string };

export interface ComponentData {
    lang: PrismLang;
    description?: string;
    snippet?: string;
    link?: Link[];
}

export interface Link {
    name?: string;
    type: string;
    url: string;
}

export interface StorybookComponentConfig {
    kind: string;
    stories?: string[];
}

export interface StyleguidistComponentConfig {
    kind: string;
}

export interface CustomUrlConfig {
    urlPath: string;
}

export interface PluginConfig {
    [key: string]: string | number | boolean;
}

export interface PluginContext {
    plugin?: PluginConfig;
    global?: GlobalConfig;
}

export type GlobalConfig = Omit<ComponentConfigFile, "projects" | "styleguides" | "link" | "components">;

export type ComponentConfig = {
    path: string;
    zeplinNames: string[];
    name?: string;
    storybook?: StorybookComponentConfig;
    styleguidist?: StyleguidistComponentConfig;
} & {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: CustomUrlConfig | any;
};

export interface ConnectPlugin {
    init?(context: PluginContext): Promise<void>;
    process(context: ComponentConfig): Promise<ComponentData>;
    supports(x: ComponentConfig): boolean;
}
