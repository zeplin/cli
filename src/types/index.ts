export * from "./link";
export * from "./prism";
