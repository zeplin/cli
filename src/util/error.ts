import { APIError, AuthError } from "../errors";

export function isAuthenticationError(err: Error): boolean {
    return APIError.isUnauthorized(err) || AuthError.isAuthError(err);
}
