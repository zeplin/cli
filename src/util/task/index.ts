export * from "./helper";
export * from "./task";
export * from "./workflow";
