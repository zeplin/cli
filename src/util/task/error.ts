import { TaskUITypes } from "./types";

export class TaskError<T> extends Error {
    ui: TaskUITypes<T>;
    cause?: Error;

    constructor(ui: TaskUITypes<T>, cause?: Error) {
        super("Task failed");
        this.ui = ui;
        this.cause = cause;
    }

    static isTaskError<T>(err: Error): err is TaskError<T> {
        return err instanceof TaskError;
    }
}
