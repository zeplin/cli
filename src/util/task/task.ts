import chalk from "chalk";
import logSymbols from "log-symbols";
import ora from "ora";
import { indent } from "../text";
import {
    State,
    TaskConstructor,
    TaskContext,
    TaskStep,
    TaskSkipFunction,
    TaskUITypes,
    TaskErrorHandler
} from "./types";
import { TaskError } from "./error";

interface RenderedUI {
    text?: string;
    subtext?: string;
}

function renderUI<T>(context: T, ui?: TaskUITypes<T>): RenderedUI {
    const msg = typeof ui === "function" ? ui(context) : ui;

    if (typeof msg === "string") {
        return { text: msg };
    } else if (typeof msg === "object") {
        const { text, subtext } = msg;
        const t = typeof text === "function" ? text(context) : text;
        const st = typeof subtext === "function" ? subtext(context) : subtext;
        return { text: t, subtext: st };
    }

    return {};
}

const defaultSkipFn = (): boolean => false;

const defaultErrorHandler = (err: Error): never => {
    const error = TaskError.isTaskError(err) ? (err.cause || err) : err;
    throw error;
};

function formatUI(renderedUI?: RenderedUI): string | undefined {
    const { text, subtext } = renderedUI || {};
    return subtext ? `${text}\n${chalk.dim(indent(subtext, "   â†’ "))}` : text;
}

function wrapStepsIntoTask<T>(steps: TaskStep<T>[]) {
    return async (ctx: T, task: Task<T>): Promise<T> => {
        for (const step of steps) {
            if (task.isPending()) {
                // eslint-disable-next-line no-await-in-loop
                const result = await step(ctx, task);
                if (result) {
                    task.render(ctx, result);
                }
            }
        }
        return ctx;
    };
}

export class Task<T = TaskContext> {
    private readonly steps: TaskStep<T>[];
    private readonly skipFn: TaskSkipFunction<T>;
    private readonly spinner: ora.Ora;
    private readonly errorHandler: TaskErrorHandler<T>;
    private ui: TaskUITypes<T>;
    private renderedUI?: RenderedUI;
    private state = State.PENDING;

    constructor(params: TaskConstructor<T>) {
        this.steps = params.steps || [];
        this.ui = params.initial || "Running task...";
        this.skipFn = params.skip || defaultSkipFn;
        this.spinner = ora(params.spinnerOptions);
        this.errorHandler = params.errorHandler || defaultErrorHandler;
    }

    isPending(): boolean {
        return this.state === State.PENDING;
    }

    isCompleted(): boolean {
        return this.state === State.COMPLETED;
    }

    isFailed(): boolean {
        return this.state === State.FAILED;
    }

    isSkipped(): boolean {
        return this.state === State.SKIPPED;
    }

    private renderUIText(context: T, ui?: TaskUITypes<T>): string | undefined {
        const renderedUI = renderUI(context, ui);
        if (renderedUI.text) {
            this.ui = ui;
            this.renderedUI = renderedUI;
        } else if (renderedUI.subtext && this.renderedUI) {
            this.renderedUI.subtext = renderedUI.subtext;
        }
        return formatUI(this.renderedUI);
    }

    /**
     * Update UI.
     *
     * Does not change current state of the task.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui New UI
     * @returns {void}
     */
    render(context: T, ui: TaskUITypes<T>): void {
        const msg = this.renderUIText(context, ui);
        if (msg) {
            this.spinner.text = msg;
            this.spinner.render();
        }
    }

    /**
     * Stops the spinner and changes it with skip symbol and persists
     *
     * Changes current state to SKIPPED.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui Persists this UI if provided.
     * @returns {void}
     */
    skip(context: T, ui?: TaskUITypes<T>): void {
        this.state = State.SKIPPED;
        this.spinner.info(this.renderUIText(context, ui));
    }

    /**
     * Stops the spinner and changes it with failure symbol and persists
     *
     * Changes current state to FAILED.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui Persists this UI if provided.
     * @returns {void}
     */
    fail(context: T, ui?: TaskUITypes<T>): void {
        this.state = State.FAILED;
        this.spinner.fail(this.renderUIText(context, ui));
    }

    /**
     * Stops the spinner and changes it with succeed symbol and persists
     *
     * Changes current state to COMPLETED.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui Persists this UI if provided.
     * @returns {void}
     */
    complete(context: T, ui?: TaskUITypes<T>): void {
        this.state = State.COMPLETED;
        this.spinner.succeed(this.renderUIText(context, ui));
    }

    isSpinning(): boolean {
        return this.spinner.isSpinning;
    }

    /**
     * Starts the spinner.
     *
     * No effect if the spinner is already started.
     *
     * Does not change current state of the task.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui Persists this UI if provided.
     * @returns {void}
     */
    startSpinner(context: T, ui?: TaskUITypes<T>): void {
        if (!this.spinner.isSpinning) {
            this.spinner.start(this.renderUIText(context, ui));
        }
    }

    /**
     * Stops the spinner.
     *
     * Persists the given UI or clears the current spinner.
     * No effect if the spinner is already stopped.
     *
     * Does not change current state of the task.
     *
     * @param {T} context Task context
     * @param {TaskUITypes<T>} ui Persists this UI if provided.
     * @returns {void}
     */
    stopSpinner(context: T, ui?: TaskUITypes<T>): void {
        if (this.spinner.isSpinning) {
            if (ui) {
                this.spinner.stopAndPersist({
                    symbol: logSymbols.info,
                    text: this.renderUIText(context, ui)
                });
            } else {
                this.spinner.stop();
                this.spinner.clear();
            }
        }
    }

    /**
     * Clears current task spinner
     *
     * @returns {void}
     */
    clearSpinner(): void {
        this.spinner.clear();
    }
    /**
     * Returns the current rendered text.
     *
     * @returns {String} Rendered text
     */
    getCurrentText(): string {
        return this.spinner.text;
    }

    async run(context?: T): Promise<T> {
        const ctx = context || Object.create(null);
        try {
            this.startSpinner(ctx, this.ui);

            const skipped = await this.skipFn(ctx);
            if (skipped) {
                if (typeof skipped === "boolean") {
                    this.skip(ctx);
                } else {
                    this.skip(ctx, skipped);
                }
            }

            if (!this.isSkipped()) {
                const task = wrapStepsIntoTask(this.steps);

                await task(ctx, this);

                if (this.isPending()) {
                    this.complete(ctx);
                }
            }

            return ctx;
        } catch (e) {
            if (TaskError.isTaskError<T>(e)) {
                e.message = this.renderUIText(ctx, e.ui) || e.message;
            }

            this.fail(ctx, e.ui);
            this.errorHandler(e, ctx, this);

            return ctx;
        }
    }
}