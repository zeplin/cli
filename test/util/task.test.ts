import { PassThrough } from "stream";
import { Task, transitionTo, Workflow } from "../../src/util/task";
import { TaskUITypes, TaskContext, TaskUI } from "../../src/util/task/types";
import { TaskError } from "../../src/util/task/error";

const ui = {
    text: "Initial static text",
    subtext: "Initial static subtext"
};

const uiFn: TaskUITypes<TaskContext> = (): TaskUI<TaskContext> => ({
    text: (): string => "Initial static text",
    subtext: (): string => "Initial static text"
});

const uiFnWithParams: TaskUITypes<TaskContext> = (ctx): TaskUI<TaskContext> => ({
    text: `Rendered text: ${ctx.param}`,
    subtext: `Rendered text: ${ctx.param}`
});

const anotherUiFnWithParams: TaskUITypes<TaskContext> = (ctx): TaskUI<TaskContext> => ({
    text: `Another rendered text: ${ctx.anotherParam}`,
    subtext: `Another rendered text: ${ctx.anotherParam}`
});

const failureUiFnWithParams: TaskUITypes<TaskContext> = (ctx): TaskUI<TaskContext> => ({
    text: `Failed rendered text: ${ctx.param}`,
    subtext: `Failed rendered text: ${ctx.param}`
});

const createMockStream = (): {
    mockedStream: PassThrough;
    output: string[];
} => {
    const mockedStream = new PassThrough();
    const output: string[] = [];

    mockedStream.on("data", (data: string) => {
        output.push(`${data}`.endsWith("\n") ? data : `${data}\n`);
    });

    return {
        mockedStream,
        output: {
            ...output,
            toString: (): string => output.join("")
        }
    };
};

describe("Task", () => {
    describe("Steps", () => {
        test("step should access the given context", async () => {
            const context = {
                randomNumber: Math.random()
            };

            const mock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    mock(ctx);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            await task.run(context);

            expect(mock).toHaveBeenCalledWith(context);
        });

        test("step should access context params set by previous steps", async () => {
            const context = {
                randomNumber: Math.random()
            };

            const newParam = Math.random();

            const mock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    ctx.newParam = newParam;
                }, (ctx): void => {
                    mock(ctx.newParam);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(mock).toHaveBeenCalledWith(newParam);
            expect(outputContext.newParam).toEqual(newParam);
        });

        test("step should not access context params set by later steps", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    mock(ctx.randomNumber);
                }, (ctx): void => {
                    ctx.randomNumber = newValue;
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(mock).toHaveBeenCalledWith(randomNumber);
            expect(outputContext.randomNumber).toEqual(newValue);
        });

        test("step should not be invoked if task is skipped at the beginning", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();

            const task = new Task({
                skip: (): boolean => true,
                steps: [(ctx): void => {
                    ctx.randomNumber = newValue;
                    mock(ctx.randomNumber);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(mock).not.toHaveBeenCalled();
            expect(outputContext.randomNumber).toEqual(randomNumber);
        });

        test("step should not be invoked if task is skipped on previous steps", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();
            const skippedMock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    ctx.randomNumber = newValue;
                    mock(newValue);
                }, (ctx, t): void => {
                    t.skip(ctx);
                }, (ctx): void => {
                    skippedMock(ctx.randomNumber);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(task.isSkipped()).toBeTruthy();
            expect(mock).toHaveBeenCalledWith(newValue);
            expect(skippedMock).not.toHaveBeenCalled();
            expect(outputContext.randomNumber).toEqual(newValue);
        });

        test("step should not be invoked if task is complete on previous steps", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();
            const skippedMock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    ctx.randomNumber = newValue;
                    mock(newValue);
                }, (ctx, t): void => {
                    t.complete(ctx);
                }, (ctx): void => {
                    skippedMock(ctx.randomNumber);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(task.isCompleted()).toBeTruthy();
            expect(mock).toHaveBeenCalledWith(newValue);
            expect(skippedMock).not.toHaveBeenCalled();
            expect(outputContext.randomNumber).toEqual(newValue);
        });

        test("step should not be invoked if task is failed on previous steps", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();
            const skippedMock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    ctx.randomNumber = newValue;
                    mock(newValue);
                }, (ctx, t): void => {
                    t.fail(ctx);
                }, (ctx): void => {
                    skippedMock(ctx.randomNumber);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(task.isFailed()).toBeTruthy();
            expect(mock).toHaveBeenCalledWith(newValue);
            expect(skippedMock).not.toHaveBeenCalled();
            expect(outputContext.randomNumber).toEqual(newValue);
        });

        test("step should not be invoked if previos step throws", async () => {
            const randomNumber = Math.random();

            const context = {
                randomNumber
            };

            const newValue = Math.random();

            const mock = jest.fn();
            const skippedMock = jest.fn();

            const task = new Task({
                steps: [(ctx): void => {
                    ctx.randomNumber = newValue;
                    mock(newValue);
                }, (ctx, t): void => {
                    t.fail(ctx);
                }, (ctx): void => {
                    skippedMock(ctx.randomNumber);
                }],
                spinnerOptions: {
                    isSilent: true
                }
            });

            const outputContext = await task.run(context);

            expect(task.isFailed()).toBeTruthy();
            expect(mock).toHaveBeenCalledWith(newValue);
            expect(skippedMock).not.toHaveBeenCalled();
            expect(outputContext.randomNumber).toEqual(newValue);
        });
    });

    describe("UI", () => {
        describe("no steps", () => {
            test("task should only render static initial UI", async () => {
                const { mockedStream, output } = createMockStream();

                const task = new Task({
                    steps: [],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(output.toString()).toMatchSnapshot();
            });

            test("task should only render function initial UI", async () => {
                const { mockedStream, output } = createMockStream();

                const task = new Task({
                    steps: [],
                    initial: uiFn,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(output.toString()).toMatchSnapshot();
            });

            test("task should only render function initial UI with params", async () => {
                const { mockedStream, output } = createMockStream();

                const task = new Task({
                    steps: [],
                    initial: uiFnWithParams,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run({
                    param: "should render this param"
                });

                expect(output.toString()).toMatchSnapshot();
            });
        });

        describe("with steps", () => {
            test("task should not update UI if not explicitly called", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [(): void => {
                        mock("invoked inside a step");
                    }],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI if render is called in a step", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [(ctx, t): void => {
                        mock("invoked inside a step");
                        ctx.param = "should render this param";
                        t.render(ctx, uiFnWithParams);
                    }],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI with transitionTo", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [
                        (ctx, t): void => {
                            mock("invoked inside a step");
                            ctx.param = "should render this param";
                            t.render(ctx, uiFnWithParams);
                        },
                        (ctx): void => {
                            ctx.anotherParam = "should render this another param";
                        },
                        transitionTo(anotherUiFnWithParams),
                        (ctx): void => {
                            ctx.anotherParam = "should render this another param 2";
                        },
                        transitionTo(anotherUiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI with step's returned value", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [
                        (): string => {
                            mock("invoked inside a step 1");
                            return "step 1 UI output";
                        },
                        (): string => {
                            mock("invoked inside a step 2");
                            return "step 2 UI output";
                        }
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(mock).toHaveBeenNthCalledWith(1, "invoked inside a step 1");
                expect(mock).toHaveBeenNthCalledWith(2, "invoked inside a step 2");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI if skip is called with another UI", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [(ctx, t): void => {
                        mock("invoked inside a step");
                        ctx.param = "should render this param";
                        t.skip(ctx, uiFnWithParams);
                    }],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(task.isSkipped()).toBeTruthy();
                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI if complete is called with another UI", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [(ctx, t): void => {
                        mock("invoked inside a step");
                        ctx.param = "should render this param";
                        t.complete(ctx, uiFnWithParams);
                    }],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(task.isCompleted()).toBeTruthy();
                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });

            test("task should update UI if fail is called with another UI", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const task = new Task({
                    steps: [(ctx, t): void => {
                        mock("invoked inside a step");
                        ctx.param = "should render this param";
                        t.fail(ctx, uiFnWithParams);
                    }],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                await task.run();

                expect(task.isFailed()).toBeTruthy();
                expect(mock).toHaveBeenCalledWith("invoked inside a step");
                expect(output.toString()).toMatchSnapshot();
            });
        });
    });

    describe("Multiple Tasks", () => {
        describe("UI", () => {
            test("tasks should run consecutively", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const random = Math.random();

                const task1 = new Task({
                    steps: [
                        (ctx): void => {
                            mock("invoked inside a task 1");
                            ctx.param = "task 1";
                            ctx.random = random;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task2 = new Task({
                    steps: [
                        (ctx): void => {
                            mock(`invoked inside a task 2, ${random}`);
                            ctx.param = "task 2";
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const tasks = new Workflow({
                    context: {},
                    tasks: [
                        task1,
                        task2
                    ]
                });

                await tasks.run();

                expect(task1.isCompleted()).toBeTruthy();
                expect(mock).toHaveBeenNthCalledWith(1, "invoked inside a task 1");

                expect(task2.isCompleted()).toBeTruthy();
                expect(mock).toHaveBeenNthCalledWith(2, `invoked inside a task 2, ${random}`);

                expect(output.toString()).toMatchSnapshot();
            });

            test("tasks should fail and throw if one of tasks throws", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();
                const notCalledMock = jest.fn();

                const random = Math.random();

                const innerError = new Error("inner error");
                const taskError = new TaskError(failureUiFnWithParams, innerError);

                const task1 = new Task({
                    steps: [
                        (ctx): void => {
                            mock("invoked inside a task 1");
                            ctx.param = "task 1";
                            ctx.random = random;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task2 = new Task({
                    steps: [
                        (): void => {
                            mock(`invoked inside a task 2, ${random}`);
                            throw taskError;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task3 = new Task({
                    steps: [
                        (ctx): void => {
                            notCalledMock(`invoked inside a task 3`);
                            ctx.param = "task 3";
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const tasks = new Workflow({
                    context: {},
                    tasks: [
                        task1,
                        task2,
                        task3
                    ]
                });

                await expect(tasks.run()).rejects.toThrowError(innerError);

                expect(task1.isCompleted()).toBeTruthy();
                expect(task2.isFailed()).toBeTruthy();
                expect(task3.isPending()).toBeTruthy();

                expect(mock).toHaveBeenNthCalledWith(1, "invoked inside a task 1");
                expect(mock).toHaveBeenNthCalledWith(2, `invoked inside a task 2, ${random}`);
                expect(notCalledMock).not.toHaveBeenCalled();

                expect(output.toString()).toMatchSnapshot();
            });

            test("tasks should fail and throw if one of tasks throws regular error", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();
                const notCalledMock = jest.fn();

                const random = Math.random();

                const error = new Error("Error text");

                const task1 = new Task({
                    steps: [
                        (ctx): void => {
                            mock("invoked inside a task 1");
                            ctx.param = "task 1";
                            ctx.random = random;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task2 = new Task({
                    steps: [
                        (): void => {
                            mock(`invoked inside a task 2, ${random}`);
                            throw error;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task3 = new Task({
                    steps: [
                        (ctx): void => {
                            notCalledMock(`invoked inside a task 3`);
                            ctx.param = "task 3";
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const tasks = new Workflow({
                    context: {},
                    tasks: [
                        task1,
                        task2,
                        task3
                    ]
                });

                await expect(tasks.run()).rejects.toThrowError(error);

                expect(task1.isCompleted()).toBeTruthy();
                expect(task2.isFailed()).toBeTruthy();
                expect(task3.isPending()).toBeTruthy();

                expect(mock).toHaveBeenNthCalledWith(1, "invoked inside a task 1");
                expect(mock).toHaveBeenNthCalledWith(2, `invoked inside a task 2, ${random}`);
                expect(notCalledMock).not.toHaveBeenCalled();

                expect(output.toString()).toMatchSnapshot();
            });

            test("tasks should run even if one of tasks throws regular error with no-op failure handler", async () => {
                const { mockedStream, output } = createMockStream();

                const mock = jest.fn();

                const random = Math.random();

                const error = new Error("Error text");

                const task1 = new Task({
                    steps: [
                        (ctx): void => {
                            mock("invoked inside a task 1");
                            ctx.param = "task 1";
                            ctx.random = random;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const task2 = new Task({
                    steps: [
                        (): void => {
                            mock(`invoked inside a task 2, ${random}`);
                            throw error;
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    },
                    errorHandler: (): void => {}
                });

                const task3 = new Task({
                    steps: [
                        (ctx): void => {
                            mock(`invoked inside a task 3`);
                            ctx.param = "task 3";
                        },
                        transitionTo(uiFnWithParams)
                    ],
                    initial: ui,
                    spinnerOptions: {
                        stream: mockedStream,
                        isEnabled: false
                    }
                });

                const tasks = new Workflow({
                    context: {},
                    tasks: [
                        task1,
                        task2,
                        task3
                    ]
                });

                await tasks.run();

                expect(task1.isCompleted()).toBeTruthy();
                expect(task2.isFailed()).toBeTruthy();
                expect(task3.isCompleted()).toBeTruthy();

                expect(mock).toHaveBeenNthCalledWith(1, "invoked inside a task 1");
                expect(mock).toHaveBeenNthCalledWith(2, `invoked inside a task 2, ${random}`);
                expect(mock).toHaveBeenNthCalledWith(3, `invoked inside a task 3`);

                expect(output.toString()).toMatchSnapshot();
            });
        });
    });
});