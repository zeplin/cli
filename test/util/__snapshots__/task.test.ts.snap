// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Multiple Tasks UI tasks should run consecutively 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[36mâ ‹[39m Rendered text: task 1
[2m   â†’ Rendered text: task 1[22m
[32mâœ”[39m Rendered text: task 1
[2m   â†’ Rendered text: task 1[22m
- Initial static text
[2m   â†’ Initial static subtext[22m
[36mâ ‹[39m Rendered text: task 2
[2m   â†’ Rendered text: task 2[22m
[32mâœ”[39m Rendered text: task 2
[2m   â†’ Rendered text: task 2[22m
"
`;

exports[`Task UI no steps task should only render function initial UI 1`] = `
"- Initial static text
[2m   â†’ Initial static text[22m
[32mâœ”[39m Initial static text
[2m   â†’ Initial static text[22m
"
`;

exports[`Task UI no steps task should only render function initial UI with params 1`] = `
"- Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
[32mâœ”[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
"
`;

exports[`Task UI no steps task should only render static initial UI 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[32mâœ”[39m Initial static text
[2m   â†’ Initial static subtext[22m
"
`;

exports[`Task UI with steps task should not update UI if not explicitly called 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[32mâœ”[39m Initial static text
[2m   â†’ Initial static subtext[22m
"
`;

exports[`Task UI with steps task should update UI if complete is called with another UI 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[32mâœ”[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
"
`;

exports[`Task UI with steps task should update UI if fail is called with another UI 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[31mâœ–[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
"
`;

exports[`Task UI with steps task should update UI if render is called in a step 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[36mâ ‹[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
[32mâœ”[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
"
`;

exports[`Task UI with steps task should update UI if skip is called with another UI 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[34mâ„¹[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
"
`;

exports[`Task UI with steps task should update UI with step's returned value 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[36mâ ‹[39m step 1 UI output
[36mâ ™[39m step 2 UI output
[32mâœ”[39m step 2 UI output
"
`;

exports[`Task UI with steps task should update UI with transitionTo 1`] = `
"- Initial static text
[2m   â†’ Initial static subtext[22m
[36mâ ‹[39m Rendered text: should render this param
[2m   â†’ Rendered text: should render this param[22m
[36mâ ™[39m Another rendered text: should render this another param
[2m   â†’ Another rendered text: should render this another param[22m
[36mâ ¹[39m Another rendered text: should render this another param 2
[2m   â†’ Another rendered text: should render this another param 2[22m
[32mâœ”[39m Another rendered text: should render this another param 2
[2m   â†’ Another rendered text: should render this another param 2[22m
"
`;
